src/MOI_wrapper.jl

```diff
@@ -0,0 +1,414 @@
+MOI.supports(::Optimizer, ::MOI.ObjectiveSense) = true
+MOI.supports(::Optimizer, ::MOI.ObjectiveFunction{SAF}) = true
+
+MOIU.supports_default_copy_to(model::Optimizer, copy_names::Bool) = !copy_names
+
+struct Param <: MOI.AbstractOptimizerAttribute
+    name::String
+end
+function MOI.set(o::Optimizer, param::Param, value)
+    o.params[param.name] = value
+    set_parameter(o.mscip, param.name, value)
+    return nothing
+end
+
+## model creation, query and modification
+
+function MOI.is_empty(o::Optimizer)
+    length(o.mscip.vars) == 0 && length(o.mscip.conss) == 0
+end
+
+function MOI.empty!(o::Optimizer)
+    # free the underlying problem
+    finalize(o.mscip)
+    # create a new one
+    o.mscip = ManagedSCIP()
+    # clear auxiliary mapping structures
+    o.index = PtrMap()
+    o.constypes = ConsTypeMap()
+    # reapply parameters
+    for pair in o.params
+        set_parameter(o.mscip, pair.first, pair.second)
+    end
+    return nothing
+end
+
+function MOI.copy_to(dest::Optimizer, src::MOI.ModelLike; kws...)
+    MOIU.automatic_copy_to(dest, src; kws...)
+end
+
+MOI.get(o::Optimizer, ::MOI.Name) = SCIPgetProbName(get_scip(o))
+MOI.set(o::Optimizer, ::MOI.Name, name::String) = @SC SCIPsetProbName(get_scip(o), name)
+
+function MOI.add_variable(o::Optimizer)
+    allow_modification(o)
+    i::Int = add_variable(o.mscip)
+    var::Ptr{SCIP_VAR} = o.mscip.vars[i][] # i == end
+    register!(o, var, i)
+    return MOI.VariableIndex(i)
+end
+
+MOI.add_variables(o::Optimizer, n) = [MOI.add_variable(o) for i=1:n]
+MOI.get(o::Optimizer, ::MOI.NumberOfVariables) = length(o.mscip.vars)
+MOI.get(o::Optimizer, ::MOI.ListOfVariableIndices) = VI.(1:length(o.mscip.vars))
+MOI.is_valid(o::Optimizer, vi::VI) = 1 <= vi.value <= length(o.mscip.vars)
+
+MOI.get(o::Optimizer, ::MOI.VariableName, vi::VI) = SCIPvarGetName(get_var(o, vi))
+function MOI.set(o::Optimizer, ::MOI.VariableName, vi::VI, name::String)
+    @SC SCIPchgVarName(get_scip(o), get_var(o, vi), name)
+end
+
+scip_vartype(::Type{BINS}) = SCIP_VARTYPE_BINARY
+scip_vartype(::Type{INTS}) = SCIP_VARTYPE_INTEGER
+function MOI.add_constraint(o::Optimizer, func::SVF, set::S) where {S <: TYPES}
+    allow_modification(o)
+    scip = get_scip(o)
+    var = get_var(o, func.variable)
+    infeasible = Ref{Ptr{SCIP_Bool}}
+    @SC SCIPchgVarType(scip, var, scip_vartype(S), infeasible[])
+    if S <: BINS
+        # need to adjust bounds for SCIP?!
+        @SC SCIPchgVarLb(scip, var, 0.0)
+        @SC SCIPchgVarUb(scip, var, 1.0)
+    end
+    # use var index for cons index of this type
+    i = func.variable.value
+    return register!(o, CI{SVF, S}(i))
+end
+
+function MOI.add_constraint(o::Optimizer, func::SVF, set::S) where S <: BOUNDS
+    allow_modification(o)
+    var = get_var(o, func.variable)
+    lb, ub = bounds(set)
+    lb == nothing || @SC SCIPchgVarLb(get_scip(o), var, lb)
+    ub == nothing || @SC SCIPchgVarUb(get_scip(o), var, ub)
+    # use var index for cons index of this type
```

Comment:
This is the right index to use (https://github.com/JuliaOpt/MathOptInterface.jl/issues/570). Is a useful error returned if a user tries to add bounds on the same variable twice?

Reply:
Not yet, see [MOI/#570](https://github.com/JuliaOpt/MathOptInterface.jl/issues/570#issuecomment-451742785)

Reply:
Consider adding a TODO to catch this case and make it an error.